### ISSUE TITLE
multiprocessing consistency across cooltools

### ISSUE THREAD

**Issue by gfudenberg** [open2c/cooltools#431]
many user-facing functions take nproc as an argument. 
e.g. 
https://github.com/open2c/cooltools/blob/c85d647239934b290c844d4db46782aff71db04d/cooltools/api/snipping.py#L824

`cooltools.sample()` does not
https://github.com/open2c/cooltools/blob/c85d647239934b290c844d4db46782aff71db04d/cooltools/api/sample.py#L54
 (however, multiprocessing is offered as an nproc argument in=cooltools.cli.random-sample does)

Any suggestion on preferred behavior?

**Comment by Phlya** [comment]
Can we make this into a decorator to ensure the argument is consistent across all tools and stays like that in the future?

**Comment by gfudenberg** [comment]
note 10/23/2023: some of the cootlools functions using pool.map should be changed to pool.imap! 

see @nvictus  comment https://github.com/open2c/cooltools/pull/464#discussion_r1369165950

**Comment by gfudenberg** [comment]
in the process of addressing this issue, @Yaoyx ran into a design choice:
(a) keep multiprocessing & factor out internal lambda functions (which pickle fails on)
(b) move to multiprocess (which uses dill instead of pickle) and otherwise keep code the same

any recommendations @golobor @nvictus @agalitsyna @sergpolly ?
an example of an internal lambda function that would need to be moved is here, inside of dots `scoring_and_histogramming()`: 

``` job = lambda tile: to_hist(to_score(tile))```
https://github.com/open2c/cooltools/blob/81b9045c58a59b374d0de5137f6c812f0a34f3ab/cooltools/api/dotfinder.py#L1307

**Comment by sergpolly** [comment]
imho multiprocess/dill is **much** more preferable, because multiprocessing/pickle prevents one more thing: https://github.com/open2c/open2c_examples/issues/29

also I've been trying to hint at that here https://github.com/open2c/cooltools/pull/477

**Comment by nvictus** [comment]
Multiprocessing will [eventually](https://github.com/python/cpython/issues/72240) let you register custom serializers, but until that happens I would stick with multiprocess+dill or another 3rd party pool manager like [loky](https://github.com/joblib/loky) or [mpire](https://github.com/sybrenjansen/mpire).

**Comment by nvictus** [comment]
We might also want to explore changing the default context (which is `"fork"` on linux) to:

```python
from multiprocess import get_context

with get_context("spawn").Pool() as pool:
    ....
````
See [this](https://pythonspeed.com/articles/python-multiprocessing/) for details.

**Comment by gfudenberg** [comment]
closed in #489
